File: es5/es5.c
#include <ctype.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

static size_t reduce(char *str, size_t len)
{
	size_t i = 0;
	while (i < len - 1) {
		if (str[i] != str[i+1] && tolower(str[i]) == tolower(str[i+1])) {
			len -= 2;
			/* copy also the terminator */
			memmove(str + i, str + i + 2, len - i + 1);
			if (i > 0)
				i -= 1;
		} else  {
			i++;
		}
	}
	return len;
}

static size_t strip(char *str, size_t len, int c)
{
	c = tolower(c);
	size_t i = 0;
	while (i < len) {
		if (tolower(str[i]) == c) {
			len--;
			memmove(str + i, str + i + 1, len - i + 1);
		} else {
			i++;
		}
	}
	return len;
}

int main(int argc, char *argv[])
{
	if (argc < 2) {
		fprintf(stderr, "Usage: %s filename\n", argv[0]);
		return -1;
	}

	FILE *input = fopen(argv[1], "rb");
	if (!input) {
		fprintf(stderr, "File %s not found\n", argv[1]);
		return -1;
	}

	char *line = NULL;
	size_t bufsize = 0;
	if (getline(&line, &bufsize, input) == -1) {
		free(line);
		fclose(input);
		return -1;
	}
	fclose(input);

	/* chop off the terminal \n */
	size_t linesize = strlen(line) - 1;
	line[linesize] = 0;

	linesize = reduce(line, linesize);
	printf("Answer1: %zu\n", linesize);

	size_t minlen = linesize;
	int chosen = -1;
	for (int c = 'a'; c <= 'z'; c++) {
		char *d = strdup(line);
		if (!d)
			return -1;

		size_t len = strip(d, linesize, c);
		len = reduce(d, len);
		if (len < minlen) {
			minlen = len;
			chosen = c;
		}
		free(d);
	}
	free(line);

	printf("Answer2: %zu by removing '%c'\n", minlen, chosen);

	return 0;
}

File: es5/es5.go
package main

import (
	"fmt"
	"io/ioutil"
	"os"
)

func lower(a byte) byte {
	return a | 32
}

func reduce(data []byte) []byte {
	i := 0
	for i < len(data)-1 {
		if data[i] != data[i+1] && lower(data[i]) == lower(data[i+1]) {
			data = append(data[:i], data[i+2:]...)
			if i > 0 {
				i--
			}
		} else {
			i++
		}
	}
	return data
}

func strip(data []byte, c byte) []byte {
	c = lower(c)
	i := 0
	for i < len(data) {
		if lower(data[i]) == c {
			data = append(data[:i], data[i+1:]...)
		} else {
			i++
		}
	}
	return data
}

func main() {
	if len(os.Args) < 2 {
		fmt.Fprintf(os.Stderr, "Usage: %s filename\n", os.Args[0])
		os.Exit(-1)
	}
	data, err := ioutil.ReadFile(os.Args[1])
	if err != nil {
		fmt.Fprintf(os.Stderr, "File %s not found\n", os.Args[1])
		os.Exit(-1)
	}

	// chop the last element (\n)
	data = data[:len(data)-1]

	data = reduce(data)
	fmt.Printf("Answer1: %d\n", len(data))

	minlen := len(data)
	var chosen byte
	for c := byte('a'); c <= byte('z'); c++ {
		dup := make([]byte, len(data))
		copy(dup, data)
		d := reduce(strip(dup, c))
		if len(d) < minlen {
			minlen = len(d)
			chosen = c
		}
	}
	fmt.Printf("Answer2: %d by removing '%c'\n", minlen, chosen)
}
